File: skills/code-architect/references/code-examples/python/flyweight.py
"""Flyweight — share intrinsic state across many objects to save memory."""

from __future__ import annotations


class Glyph:
    """Stores only intrinsic (shared) state: the character's shape."""

    def __init__(self, symbol: str) -> None:
        self.symbol = symbol

    def render(self, x: int, y: int) -> str:
        # Extrinsic state (position) is passed in by the caller.
        return f"'{self.symbol}' @ ({x},{y})"


class GlyphFactory:
    """Returns shared Glyph instances, creating each symbol only once."""

    def __init__(self) -> None:
        self._cache: dict[str, Glyph] = {}

    def get(self, symbol: str) -> Glyph:
        if symbol not in self._cache:
            self._cache[symbol] = Glyph(symbol)
        return self._cache[symbol]

    def count(self) -> int:
        return len(self._cache)


if __name__ == "__main__":
    factory = GlyphFactory()
    text = "abba"
    for i, ch in enumerate(text):
        print(factory.get(ch).render(i, 0))
    print(f"distinct glyph objects for {len(text)} chars: {factory.count()}")

File: skills/code-architect/references/code-examples/java/Flyweight.java
// Flyweight — share intrinsic state across many objects to save memory.

import java.util.HashMap;
import java.util.Map;

// Intrinsic, shareable state.
class Glyph {
    private final char symbol;
    private final String font;
    Glyph(char symbol, String font) { this.symbol = symbol; this.font = font; }
    String render(int x, int y) { // extrinsic state passed in
        return symbol + "@" + x + "," + y + "(" + font + ")";
    }
}

class GlyphFactory {
    private final Map<String, Glyph> pool = new HashMap<>();
    Glyph get(char symbol, String font) {
        String key = symbol + ":" + font;
        return pool.computeIfAbsent(key, k -> new Glyph(symbol, font));
    }
    int distinctCount() { return pool.size(); }
}

public class Flyweight {
    public static void main(String[] args) {
        GlyphFactory factory = new GlyphFactory();
        String text = "aaab";
        int x = 0;
        StringBuilder out = new StringBuilder();
        for (char c : text.toCharArray()) {
            Glyph g = factory.get(c, "mono");
            out.append(g.render(x++, 0)).append(" ");
        }
        System.out.println(out.toString().trim());
        System.out.println("shared glyphs: " + factory.distinctCount());
    }
}
